File: 2024/07.py
#!/usr/bin/env python

import sys


def test(target: int, i: int, nums: [int]):
    if i == 0:
        return target == nums[0]

    if target % nums[i] == 0 and test(int(target / nums[i]), i - 1, nums):
        return True
    if target >= nums[i] and test(target - nums[i], i - 1, nums):
        return True
    return False


def test2(target: int, i: int, nums: [int]):
    if i == 0:
        return target == nums[i]

    if target % nums[i] == 0 and test2(int(target / nums[i]), i - 1, nums):
        return True
    if target >= nums[i] and test2(target - nums[i], i - 1, nums):
        return True
    try:
        sn = str(nums[i])
        st = str(target)
        if st.endswith(sn) and test2(int(str(target)[:-len(sn)]), i - 1, nums):
            return True
    except Exception:
        pass
    return False


task1 = 0
task2 = 0
for line in sys.stdin.readlines():
    vals = line.split()
    target = int(vals[0][:-1])
    nums = [int(i) for i in vals[1:]]
    if test(target, len(nums) - 1, nums):
        task1 += target
    if test2(target, len(nums) - 1, nums):
        task2 += target

print(task1)
print(task2)

File: 2024/07.cpp
#include "../test.hpp"
#include <iterator>

namespace {

using namespace boost::ut;

using IntT = int64_t;
using NumsT = std::vector<IntT>;

bool Check(IntT target, int i, NumsT nums)
{
    if (!i)
        return target == nums[0];
    if (target % nums[i] == 0 && Check(target / nums[i], i - 1, nums))
        return true;
    if (target >= nums[i] && Check(target - nums[i], i - 1, nums))
        return true;
    return false;
}

bool Check2(IntT target, int i, NumsT nums)
{
    if (!i)
        return target == nums[0];
    if (target % nums[i] == 0 && Check2(target / nums[i], i - 1, nums))
        return true;
    if (target >= nums[i] && Check2(target - nums[i], i - 1, nums))
        return true;
    std::string st = std::to_string(target);
    std::string sn = std::to_string(nums[i]);
    if (st.ends_with(sn) && st.size() > sn.size()) {
        auto nt = st.substr(0, st.size() - sn.size());
        if (Check2(std::stoll(nt), i - 1, nums))
            return true;
    }
    return false;
}

std::pair<IntT, IntT> Task(std::istream &&is)
{
    IntT task1{}, task2{};

    std::string line;
    while (getline(is, line)) {
        std::stringstream iss{line};
        IntT target{};
        iss >> target;
        iss.get();
        NumsT nums;
        std::copy(std::istream_iterator<int>(iss), std::istream_iterator<int>{}, std::back_inserter(nums));
        if (Check(target, nums.size() - 1, nums))
            task1 += target;
        if (Check2(target, nums.size() - 1, nums))
            task2 += target;
    }

    return {task1, task2};
}

suite s = [] {
    "07"_test = [] {
        const char *const TEST1 = R"(190: 10 19
3267: 81 40 27
83: 17 5
156: 15 6
7290: 6 8 6 15
161011: 16 10 13
192: 17 8 14
21037: 9 7 18 13
292: 11 6 16 20)";
        auto test = Task(std::istringstream{TEST1});
        expect(3749_i == test.first);
        expect(11387_i == test.second);

        auto task = Task(std::ifstream{INPUT});
        Printer::Print(__FILE__, "1", task.first);
        Printer::Print(__FILE__, "2", task.second);
    };
};

} //namespace;
